lib/puppet/pops/loader/puppet_function_instantiator.rb

```diff
@@ -43,7 +43,8 @@ def self.create(loader, typed_name, source_ref, pp_code_string)
     # Adapt the function definition with loader - this is used from logic contained in it body to find the
     # loader to use when making calls to the new function API. Such logic have a hard time finding the closure (where
     # the loader is known - hence this mechanism
-    Puppet::Pops::Adapters::LoaderAdapter.adapt(the_function_definition).loader = loader
+    private_loader = loader.private_loader
+    Puppet::Pops::Adapters::LoaderAdapter.adapt(the_function_definition).loader = private_loader
```

Comment:
that sounds like a problem as well... need to read up on how the dependent loaders are organized; module A depending on module B, should see module B's public loader, but not its private. That is how it (should be) designed to work.